rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,415 @@
+
+    // Update the total number of allocated blocks.
+    let old_total_blocks = {
+	let c = meta_data::total_allocated_blocks::get();
+	let c_ = c + inc_block_count as u64;
+	// to do -- assert c_ is less than meta_data::max::BLOCKS
+	meta_data::total_allocated_blocks::set(c_);
+	c
+    };
+
+    // Update this region's page count, in both places where we record it (heap object, region table).
+    {
+	let r_id = RegionId::from_id((*r).id);
+	let c = meta_data::region_table::get(&r_id);
+
+	// Region table agrees with heap object's field.
+	assert_eq!(c, Some(RegionSizeInPages(old_page_count.into())));
+
+	// Increase both:
+	(*r).page_count += new_pages_;
+	meta_data::region_table::set(&r_id, Some(RegionSizeInPages((*r).page_count.into())));
+    }
+
+    let new_vec_pages = alloc_blob(mem, Bytes(new_block_count * 2));
+    let old_vec_byte_count = old_block_count * 2;
+
+    let new_pages = AccessVector::from_value(&new_vec_pages);
+    let old_pages = AccessVector::from_value(&(*r).vec_pages);
+
+    // Copy old region-block associations into new heap object.
+    for i in 0..old_vec_byte_count {
+	new_pages.0.set(i, old_pages.0.get(i));
+    }
+
+    println!(80, "region_grow id={} (old_block_count, new_block_count) = ({}, {})", (*r).id, old_block_count, new_block_count);
+
+    // Record new associations, between the region and each new block:
+    // - in block_region_table (stable memory, for persistence).
+    // - in region representation (heap memory, for fast access operations).
+    for i in old_block_count..new_block_count {
+	// rel_i starts at zero.
+	let rel_i : u16 = (i - old_block_count as u32) as u16;
+
+	// (to do -- handle case where allocating this way has run out.)
+	let block_id : u16 = (old_total_blocks + rel_i as u64) as u16;
+
+	println!(50, "  region_grow id={} (slot index) i={} block_id={}", (*r).id, i, block_id);
+
+	// Update stable memory with new association.
+	let assoc =  Some((RegionId::from_id((*r).id), i as u16));
+	meta_data::block_region_table::set(BlockId(block_id), assoc.clone());
+
+	if true { // temp sanity testing: read back the data we just wrote.
+	    let assoc_ = meta_data::block_region_table::get(BlockId(block_id));
+	    assert_eq!(assoc, assoc_)
+	}
+
+	new_pages.set_ith_block_id(i, &BlockId(block_id));
+
+	if true { // temp sanity testing: read back the data we just wrote.
+	    let block_id_ = new_pages.get_ith_block_id(i);
+	    assert_eq!(BlockId(block_id), block_id_);
+	}
+    }
+
+    (*r).vec_pages = new_vec_pages;
+    old_page_count.into()
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_load_byte<M: Memory>(_mem: &mut M, r: Value, offset: u64) -> u8 {
+    let r = RegionObject::from_value(&r);
+    let offset = r.local_offset_into_global_offset(offset);
+    let mut dst : [u8; 1] = [0];
+    crate::ic0_stable::nicer::read(offset, &mut dst);
+    dst[0]
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_store_byte<M: Memory>(_mem: &mut M, r: Value, offset: u64, byte: u8) {
```

Comment:
As you suggest, I followed the pattern that was there for `stableMemoryLoadNat8` and `stableMemoryStoreNat8` used, where we use `TaggedSmallWord.msb_adjust` after and `TaggedSmallWord.lsb_adjust` before the call, respectively.  My test seems to confirm that this is working right now.

Reply:
>If these functions are meant to be called from compile.ml then the u8 result/arg should either be a Value or u32 depending on where you want the bit-shifting to encode the byte as a value to take place. I'd say use u32 (with byte in low bits) and do the most efficient thing on the other (ML) side.
NIT: I think the Rust signatures for load/store should take the byte as a u32 value (in low byte) - since that is the interface presented to the Ocaml code - there's no hard guarantee that Rust will use a 32-bit word for the byte otherwise - although it almost certainly does.